46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+highest severity by default because otherwise some events wont be caught is a
+irrecoverable failure is triggered before.
+
+## Design Details
+We keep the current model of the states, but modify the behavior of the failure
+states. If we are in a failure state that is recoverable the polling of
+the agent is stopped otherwise the agent is still added back for normal polling.
+
+A new failure object will be introduced which holds the all the revocation
+events that are being produced by the checks and if any of the failures makes
+the agent not recoverable. Irrecoverable events currently are retry timeouts and
+quote validation.
+
+All parts of the validation process which such as (`check_qoute` and
+`check_pcrs`) will append all events to that object instead of returning false.
+If the validation process advances without a step that failed the failure object
+must be marked as irrecoverable and only then the function can return without
+validating further. The state of the agent after that should be
+`QUOTE_FAILED_IRRECOVERABLE`.
+
+The highest severity is 0 and the higher the number the number the lower the
+severity. This makes it easy to add less sever levels without the need to
+increase the level of the highest severity.
+
+A new table `revocation_events` is introduced to save the already sent
+notifications. It contains the columns:
+  * `agent_id`: The id of agent where the event was created.
+  * `event_id`: The event ID string
+  * `severity_level`: The severity of the event.
+  * `context`: String or JSON object that contains more information about that event.
+
+The `process_agent` function gets a new argument that can contain a failure
+object. If the status is `QUOTE_FAILED` the events from failure object are
+checked against the database and if there are occurring the first time a message
+is send and they are added to the database. Otherwise the event is ignored.
+
+In the verifier API for the agent a new field is exposed that contains the list
+of the revocation events, with their respected level and context.
+
+To the send revocation message two new field are added: `severity_level` and
+`context`. Such that the other agents can act accordingly.
+
+(To reduce bandwidth all the generated messages could also be bundled into one,
+but that would introduce more complexity of parsing the messages for the
+recipient.)
+
+If a agent gets removed from the verifier then all entries in
+`revocation_events` for the agent will be removed. (An option to clear the old
+events from a agent could be introduced, but a reset can already implemented by
+removing and then adding the agent to the verifier)
+
+### Events
+Events are tagged with an event id. Which has the following schema
+`component.[sub_component].event`.
+
+* `component`: Name of the part of Keylime where the event was generated.
+  * Current would be: `quote_validation`, `pcr_validation`, `measured_boot`,
```

Comment:
This is a really really really good idea, and IMO should be implemented first (probably with an extra column on the main agent table) in an self-contained manner.